### PR TITLE
fixed issue 177

### LINKED ISSUE #177: Fail to push if port 443 is explicitly specified in the hostname
If the port `443` is explicitly set to the hostname like `myregistry:443/myrepo`, blob push will fail since the hostname of the `Location` header in the response is set to `myregistry` instead of `myregistry:443`.

https://github.com/oras-project/oras-go/blob/4ed4d30e890b9ff39c8dc231f98e80637badce00/registry/remote/repository.go#L520

Thus, the auth client missed the cache and fails.

### PR DESCRIPTION
**Pull request by patrickzheng200**
Signed-off-by: Patrick Zheng <patrickzheng@microsoft.com>
Fixed #177

### CODE REVIEW

**Comment by patrickzheng200** [comment]
@Wwwsylvia PR needs review. Thanks.

**Comment by Wwwsylvia** [comment]
@patrickzheng200 Some unit tests failed. Can you take a look?

**Comment by patrickzheng200** [comment]
> @patrickzheng200 Some unit tests failed. Can you take a look?

Taking a look.

File: registry/remote/repository_test.go [hunk 1]
```diff
@@ -2203,3 +2205,53 @@ func Test_ManifestStore_FetchReference(t *testing.T) {
 		t.Errorf("Manifests.FetchReference() = %v, want %v", got, manifest)
 	}
 }
+
+// Testing fix of Issue 177 with dummy client credential
+func Test_BlobStore_Push_Fetch_Iss177(t *testing.T) {
```
Comment (Wwwsylvia) [inline registry/remote/repository_test.go:2210]: We should call out the testing scenario in the function name, such as `Test_BlobStore_Push_Port443` or something like that.

File: registry/remote/repository.go [hunk 2]
```diff
@@ -506,6 +506,7 @@ func (s *blobStore) Push(ctx context.Context, expected ocispec.Descriptor, conte
 	if err != nil {
 		return err
 	}
+	hs := strings.Split(req.URL.Host, ":")
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:509]: The name `hs` is confusing. We need a more descriptive name as this variable is used in 20 lines later. How about `reqHost`?

File: registry/remote/repository.go [hunk 3]
```diff
@@ -523,6 +524,10 @@ func (s *blobStore) Push(ctx context.Context, expected ocispec.Descriptor, conte
 	if err != nil {
 		return err
 	}
+	// work-around solution for https://github.com/oras-project/oras-go/issues/177
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:527]: We need a more detailed purpose here.

**Review by shizhMSFT — COMMENTED** [review]
Please add unit test for issue #177, and cover at least the following test cases:

Test case 1:
- Host: registry.wabbit-networks.io:443
- Location: registry.wabbit-networks.io
- PlainHTTP: true

Test case 2:
- Host: registry.wabbit-networks.io:443
- Location: myregistry.com
- PlainHTTP: false

Test case 3:
- Host: registry.wabbit-networks.io:443
- Location: registry.wabbit-networks.io:443
- PlainHTTP: false

Test case 4:
- Host: myregistry.com
- Location: registry.wabbit-networks.io:443
- PlainHTTP: false

Test case 5:
- Host: registry.wabbit-networks.io
- Location: registry.wabbit-networks.io
- PlainHTTP: false

Test case 6:
- Host: registry.wabbit-networks.io
- Location: registry.wabbit-networks.io
- PlainHTTP: true

File: registry/remote/repository.go (hunk 3 above)
Comment (patrickzheng200) [reply registry/remote/repository.go:528]: Fix updated

File: registry/remote/repository.go [hunk 4]
```diff
@@ -523,6 +525,16 @@ func (s *blobStore) Push(ctx context.Context, expected ocispec.Descriptor, conte
...
+	// For some registries, if the port 443 is explicitly set to the hostname like myregistry:443/myrepo,
+	// blob push will fail since the hostname of the Location header in the response is set to myregistry
+	// instead of myregistry:443. For example: nv2.azurewebsites.net:443 becomes nv2.azurewebsites.net
```
Comment (shizhMSFT) [inline registry/remote/repository.go:531]: Can you format the comment block to have a width of 80?

Comment (shizhMSFT) [inline registry/remote/repository.go:531]: Please use registries with neutral names for the examples.

File: registry/remote/repository_test.go [hunk 5]
```diff
@@ -2203,3 +2206,267 @@ func Test_ManifestStore_FetchReference(t *testing.T) {
...
+
+	err = store.Push(ctx, blobDesc, bytes.NewReader(blob))
+	if err != nil {
+		t.Fatalf("Blobs.Push() error = %v", err)
+	}
+}
+
+// Helper function to create a registry.BlobStore for
+// Test_BlobStore_Push_Port443_HTTPS
+func BlobStore_Push_Port443_HTTPS_create_store(uri *url.URL, testRegistry string) (registry.BlobStore, error) {
```
Comment (Wwwsylvia) [inline registry/remote/repository_test.go:2327]: Same here.

File: registry/remote/repository_test.go [hunk 6]
```diff
@@ -2203,3 +2206,267 @@ func Test_ManifestStore_FetchReference(t *testing.T) {
...
+	if err != nil {
+		return nil, err
+	}
+	resp.Request.Host = t.mockHost
+	resp.Request.URL.Host = t.mockHost
+	return resp, nil
+}
+
+// Helper function to create a registry.BlobStore for Test_BlobStore_Push_Port443
+func BlobStore_Push_Port443_create_store(uri *url.URL, testRegistry string) (registry.BlobStore, error) {
```
Comment (Wwwsylvia) [inline registry/remote/repository_test.go:2239]: Helper functions should be unexported.

File: registry/remote/repository_test.go (hunk 5 above)
Comment (patrickzheng200) [reply registry/remote/repository_test.go]: updated

File: registry/remote/repository_test.go (hunk 6 above)
Comment (patrickzheng200) [reply registry/remote/repository_test.go]: updated

File: registry/remote/repository_test.go [hunk 7]
```diff
@@ -2203,3 +2206,269 @@ func Test_ManifestStore_FetchReference(t *testing.T) {
...
+	defer ts.Close()
+	uri, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("invalid test https server: %v", err)
+	}
+
+	ctx := context.Background()
+	// Test case with Host: "registry.wabbit-networks.io:443",
+	// Location: "registry.wabbit-networks.io"
+	testReistry := "registry.wabbit-networks.io:443"
```
Comment (Wwwsylvia) [inline registry/remote/repository_test.go:2396]: nit: typo
```suggestion
	testRegistry := "registry.wabbit-networks.io:443"
```

**Review by Wwwsylvia — APPROVED** [review]
LGTM

File: registry/remote/repository_test.go (hunk 7 above)
Comment (patrickzheng200) [reply registry/remote/repository_test.go]: Typo fixed

**Review by Wwwsylvia — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**